zerver/tornado/event_queue.py

```diff
@@ -1272,28 +1276,28 @@ def process_presence_event(event: Mapping[str, Any], users: Iterable[int]) -> No
         # since presence events are pretty ephemeral in nature.
         logging.warning("Dropping some obsolete presence events after upgrade.")
 
-    slim_event = dict(
+    legacy_event = dict(
         type="presence",
         user_id=event["user_id"],
         server_timestamp=event["server_timestamp"],
-        presence=event["presence"],
+        presence=event["legacy_presence"],
     )
 
-    legacy_event = dict(
+    modern_event = dict(
         type="presence",
-        user_id=event["user_id"],
-        email=event["email"],
-        server_timestamp=event["server_timestamp"],
-        presence=event["presence"],
+        presences={event["user_id"]: event["modern_presence"]},
     )
 
     for user_profile_id in users:
         for client in get_client_descriptors_for_user(user_profile_id):
             if client.accepts_event(event):
-                if client.slim_presence:
-                    client.add_event(slim_event)
+                if client.modern_presence_format:
+                    presence_event = modern_event.copy()
                 else:
-                    client.add_event(legacy_event)
+                    presence_event = legacy_event.copy()
```

Comment:
Doing all these copies is expensive; I think you can just make `slim_event` too; if I'm not mistaken, `slim_presence` only affects the legacy format, right?